README.rst
===============================
LMS-Insights Data Interface
===============================


.. image:: https://img.shields.io/pypi/v/insights_data.svg
        :target: https://pypi.python.org/pypi/insights_data

.. image:: https://img.shields.io/travis/doctoryes/insights_data.svg
        :target: https://travis-ci.org/doctoryes/insights_data

.. image:: https://readthedocs.org/projects/insights-data/badge/?version=latest
        :target: https://insights-data.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/doctoryes/insights_data/shield.svg
     :target: https://pyup.io/repos/github/doctoryes/insights_data/
     :alt: Updates


Python module to use in querying Insights data from the LMS.


* Free software: Apache Software License 2.0
* Documentation: https://insights-data.readthedocs.io.


Features
--------

* TODO

Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage



HISTORY.rst
=======
History
=======

0.1.0 (2017-04-13)
------------------

* First release on PyPI.


tests/test_insights_data.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_insights_data
----------------------------------

Tests for `insights_data` module.
"""

import pytest


from insights_data import insights_data


@pytest.fixture
def response():
    """Sample pytest fixture.
    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument.
    """
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'doctoryes/insights_data'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


insights_data/insights_data.py
# -*- coding: utf-8 -*-
import re
import requests


class APIKeyAuth(requests.auth.AuthBase):
    def __init__(self, api_key):
        self.api_key = api_key
    def __call__(self, req):
        # Add the API key to the headers.
        req.headers['Authorization'] = 'Token {}'.format(self.api_key)
        return req


class LMSInsightsClient(object):
    """
    Class used to query Insights API data from the LMS.
    """
    def __init__(self, base_insights_api_url, api_key=None):
        # Strip trailing slash if exists.
        self.api_base_url = re.sub(r'/$', '', base_insights_api_url)
        self.api_key = api_key

    def _get_api_courses_results(self, course_id, by_filter=None):
        """
        Utility method for retrieving course info.
        """
        url = '{base_url}/courses/{course_id}/enrollment/'.format(
            base_url=self.api_base_url,
            course_id=course_id
        )
        if by_filter:
            url += by_filter + '/'
        results = requests.get(url, auth=APIKeyAuth(self.api_key))
        return results.json()

    def get_current_course_enrollment(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data.

        Returns:
            Dictionary in this form:
              {
                "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                "date": "2017-03-26",
                "count": 236107,
                "created": "2017-03-28T012204"
              }
        """
        return self._get_api_courses_results(course_id)[0]

    def get_current_course_enrollment_by_mode(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data by mode.

        Returns:
            Dictionary in this form:
              {
                "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                "date": "2017-03-26",
                "count": 236107,
                "cumulative_count": 269777,
                "created": "2017-03-28T010109",
                "audit": 11915,
                "credit": 0,
                "honor": 222320,
                "professional": 0,
                "verified": 1872
              }
        """
        return self._get_api_courses_results(course_id, 'mode')[0]

    def get_current_course_enrollment_by_birth_year(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data by birth year.

        Returns:
            List in this form:
                [
                  ...,
                  {
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "date": "2017-03-26",
                    "birth_year": 1984,
                    "count": 6698,
                    "created": "2017-03-28T014331"
                  },
                  {
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "date": "2017-03-26",
                    "birth_year": 1985,
                    "count": 7259,
                    "created": "2017-03-28T014333"
                  },
                  ...
                ]
        """
        return self._get_api_courses_results(course_id, 'birth_year')

    def get_current_course_enrollment_by_education(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data by education level.

        Returns:
            List in this form:
                [
                  ...,
                  {
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "date": "2017-03-26",
                    "education_level": "doctorate",
                    "count": 4820,
                    "created": "2017-03-28T013248"
                  },
                  {
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "date": "2017-03-26",
                    "education_level": "junior_secondary",
                    "count": 4101,
                    "created": "2017-03-28T013249"
                  },
                  ...
                ]
        """
        return self._get_api_courses_results(course_id, 'education')

    def get_current_course_enrollment_by_location(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data by student location.

        Returns:
            List in this form:
                [
                  ...,
                  {
                    "date": "2017-01-19",
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "country": {
                      "alpha2": "VE",
                      "alpha3": "VEN",
                      "name": "Venezuela"
                    },
                    "count": 612,
                    "created": "2017-01-19T231938"
                  },
                  {
                    "date": "2017-01-19",
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "country": {
                      "alpha2": "VG",
                      "alpha3": "VGB",
                      "name": "Virgin Islands (British)"
                    },
                    "count": 4,
                    "created": "2017-01-19T231939"
                  },
                  ...
                ]
        """
        return self._get_api_courses_results(course_id, 'location')


insights_data/__init__.py
# -*- coding: utf-8 -*-
import re
import requests

__author__ = """John Eskew"""
__email__ = 'jeskew@edx.org'
__version__ = '0.1.0'


class APIKeyAuth(requests.auth.AuthBase):
    def __init__(self, api_key):
        self.api_key = api_key
    def __call__(self, req):
        # Add the API key to the headers.
        req.headers['Authorization'] = 'Token {}'.format(self.api_key)
        return req


class LMSInsightsClient(object):
    """
    Class used to query Insights API data from the LMS.
    """
    def __init__(self, base_insights_api_url, api_key=None):
        # Strip trailing slash if exists.
        self.api_base_url = re.sub(r'/$', '', base_insights_api_url)
        self.api_key = api_key

    def _get_api_courses_results(self, course_id, by_filter=None):
        """
        Utility method for retrieving course info.
        """
        url = '{base_url}/courses/{course_id}/enrollment/'.format(
            base_url=self.api_base_url,
            course_id=course_id
        )
        if by_filter:
            url += by_filter + '/'
        results = requests.get(url, auth=APIKeyAuth(self.api_key))
        return results.json()

    def get_current_course_enrollment(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data.

        Returns:
            Dictionary in this form:
              {
                "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                "date": "2017-03-26",
                "count": 236107,
                "created": "2017-03-28T012204"
              }
        """
        return self._get_api_courses_results(course_id)[0]

    def get_current_course_enrollment_by_mode(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data by mode.

        Returns:
            Dictionary in this form:
              {
                "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                "date": "2017-03-26",
                "count": 236107,
                "cumulative_count": 269777,
                "created": "2017-03-28T010109",
                "audit": 11915,
                "credit": 0,
                "honor": 222320,
                "professional": 0,
                "verified": 1872
              }
        """
        return self._get_api_courses_results(course_id, 'mode')[0]

    def get_current_course_enrollment_by_birth_year(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data by birth year.

        Returns:
            List in this form:
                [
                  ...,
                  {
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "date": "2017-03-26",
                    "birth_year": 1984,
                    "count": 6698,
                    "created": "2017-03-28T014331"
                  },
                  {
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "date": "2017-03-26",
                    "birth_year": 1985,
                    "count": 7259,
                    "created": "2017-03-28T014333"
                  },
                  ...
                ]
        """
        return self._get_api_courses_results(course_id, 'birth_year')

    def get_current_course_enrollment_by_education(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data by education level.

        Returns:
            List in this form:
                [
                  ...,
                  {
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "date": "2017-03-26",
                    "education_level": "doctorate",
                    "count": 4820,
                    "created": "2017-03-28T013248"
                  },
                  {
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "date": "2017-03-26",
                    "education_level": "junior_secondary",
                    "count": 4101,
                    "created": "2017-03-28T013249"
                  },
                  ...
                ]
        """
        return self._get_api_courses_results(course_id, 'education')

    def get_current_course_enrollment_by_location(self, course_id):
        """
        Params:
            course_id (str): Course ID for which to retrieve enrollment data by student location.

        Returns:
            List in this form:
                [
                  ...,
                  {
                    "date": "2017-01-19",
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "country": {
                      "alpha2": "VE",
                      "alpha3": "VEN",
                      "name": "Venezuela"
                    },
                    "count": 612,
                    "created": "2017-01-19T231938"
                  },
                  {
                    "date": "2017-01-19",
                    "course_id": "course-v1:LinuxFoundationX+LFS101x.2+1T2015",
                    "country": {
                      "alpha2": "VG",
                      "alpha3": "VGB",
                      "name": "Virgin Islands (British)"
                    },
                    "count": 4,
                    "created": "2017-01-19T231939"
                  },
                  ...
                ]
        """
        return self._get_api_courses_results(course_id, 'location')
